plugin/cache/cache.go

```diff
@@ -105,7 +106,40 @@ type ResponseWriter struct {
 	state  request.Request
 	server string // Server handling the request.
 
-	prefetch bool // When true write nothing back to the client.
+	prefetch   bool // When true write nothing back to the client.
+	remoteAddr net.Addr
+}
+
+// newPrefetchResponseWriter returns a Cache ResponseWriter to be used in
+// prefetch requests. It ensures RemoteAddr() can be called even after the
+// original connetion has already been closed.
+func newPrefetchResponseWriter(server string, state request.Request, c *Cache) *ResponseWriter {
```

Comment:
I think it is better to have a new type PrefetchResponseWriter that does the prefetch, the prefetch bool can then also be dropped from the original ResponseWriter.

I looked at bit; the only downside is that WriteMsg needs to be duplicated. WDYT?

Reply:
I'd love to refactor plugin/cache at some point and decouple responsibilities. The prefetch bool is a great indicator that multiple concerns are handled by a single type. There is more to it though, the `Set()` method should really be defined on cache and not the responsewriter.

At the moment my main goal is to get a stable version deployed, so refactoring the existing package structure is a second priority. I'm happy to look into it end of the week.